File: modules/core/misc/java/test/Point3Test.java
package org.opencv.test.core;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;
import org.opencv.core.Point;
import org.opencv.core.Point3;
import org.opencv.test.OpenCVTestCase;

public class Point3Test extends OpenCVTestCase {

    private Point3 p1;
    private Point3 p2;

    @Override
    public void setUp() throws Exception {
        super.setUp();

        p1 = new Point3(2, 2, 2);
        p2 = new Point3(1, 1, 1);
    }

    @Test
    public void testClone() {
        Point3 truth = new Point3(1, 1, 1);
        p1 = truth.clone();
        assertEquals(truth, p1);
    }

    @Test
    public void testCross() {
        Point3 dstPoint = p1.cross(p2);
        Point3 truth = new Point3(0, 0, 0);
        assertEquals(truth, dstPoint);
    }

    @Test
    public void testDot() {
        double result = p1.dot(p2);
        assertEquals(6.0, result, 0);
    }

    @Test
    public void testEqualsObject() {
        boolean flag = p1.equals(p1);
        assertTrue(flag);

        flag = p1.equals(p2);
        assertFalse(flag);
    }

    @Test
    public void testHashCode() {
        assertEquals(p1.hashCode(), p1.hashCode());
    }

    @Test
    public void testPoint3() {
        p1 = new Point3();

        assertNotNull(p1);
        assertTrue(0 == p1.x);
        assertTrue(0 == p1.y);
        assertTrue(0 == p1.z);
    }

    @Test
    public void testPoint3DoubleArray() {
        double[] vals = { 1, 2, 3 };
        p1 = new Point3(vals);

        assertTrue(1 == p1.x);
        assertTrue(2 == p1.y);
        assertTrue(3 == p1.z);
    }

    @Test
    public void testPoint3DoubleDoubleDouble() {
        p1 = new Point3(1, 2, 3);

        assertEquals(1., p1.x, 0);
        assertEquals(2., p1.y, 0);
        assertEquals(3., p1.z, 0);
    }

    @Test
    public void testPoint3Point() {
        Point p = new Point(2, 3);
        p1 = new Point3(p);

        assertEquals(2., p1.x, 0);
        assertEquals(3., p1.y, 0);
        assertEquals(0., p1.z, 0);
    }

    @Test
    public void testSet() {
        double[] vals1 = {};
        p1.set(vals1);

        assertEquals(0., p1.x, 0);
        assertEquals(0., p1.y, 0);
        assertEquals(0., p1.z, 0);

        double[] vals2 = { 3, 6, 10 };
        p1.set(vals2);

        assertEquals(3., p1.x, 0);
        assertEquals(6., p1.y, 0);
        assertEquals(10., p1.z, 0);
    }

    @Test
    public void testToString() {
        String actual = p1.toString();
        String expected = "{2.0, 2.0, 2.0}";
        assertEquals(expected, actual);
    }

}

File: modules/core/misc/objc/test/Point3Test.swift
//
//  Point3Test.swift
//
//  Created by Giles Payne on 2020/01/31.
//

import XCTest
import OpenCV

class Point3Test: OpenCVTestCase {

    let p1 = Point3i(x: 2, y: 2, z: 2)
    let p2 = Point3i(x: 1, y: 1, z: 1)

    func testClone() {
        let truth = Point3i(x: 1, y: 1, z: 1)
        let p1 = truth.clone()
        XCTAssertEqual(truth, p1)
    }

    func testCross() {
        let dstPoint = p1.cross(p2)
        let truth = Point3i(x: 0, y: 0, z: 0)
        XCTAssertEqual(truth, dstPoint)
    }

    func testDot() {
        let result = p1.dot(p2)
        XCTAssertEqual(6.0, result)
    }

    func testEqualsObject() {
        var flag = p1 == p1
        XCTAssert(flag)

        flag = p1 == p2
        XCTAssertFalse(flag)
    }

    func testHashCode() {
        XCTAssertEqual(p1.hash(), p1.hash())
    }

    func testPoint3() {
        let p1 = Point3i()

        XCTAssertNotNil(p1)
        XCTAssert(0 == p1.x)
        XCTAssert(0 == p1.y)
        XCTAssert(0 == p1.z)
    }

    func testPoint3DoubleArray() {
        let vals:[Double] = [1, 2, 3]
        let p1 = Point3i(vals: vals as [NSNumber])

        XCTAssert(1 == p1.x)
        XCTAssert(2 == p1.y)
        XCTAssert(3 == p1.z)
    }

    func testPoint3DoubleDoubleDouble() {
        let p1 = Point3i(x: 1, y: 2, z: 3)

        XCTAssertEqual(1, p1.x)
        XCTAssertEqual(2, p1.y)
        XCTAssertEqual(3, p1.z)
    }

    func testPoint3Point() {
        let p = Point(x: 2, y: 3)
        let p1 = Point3i(point: p)

        XCTAssertEqual(2, p1.x)
        XCTAssertEqual(3, p1.y)
        XCTAssertEqual(0, p1.z)
    }

    func testSet() {
        let vals1:[Double] = []
        p1.set(vals: vals1 as [NSNumber]);

        XCTAssertEqual(0, p1.x)
        XCTAssertEqual(0, p1.y)
        XCTAssertEqual(0, p1.z)

        let vals2 = [3, 6, 10]
        p1.set(vals: vals2 as [NSNumber])

        XCTAssertEqual(3, p1.x)
        XCTAssertEqual(6, p1.y)
        XCTAssertEqual(10, p1.z)
    }

    func testToString() {
        let actual = "\(p1)"
        let expected = "Point3i {2,2,2}"
        XCTAssertEqual(expected, actual)
    }

}
